base/applications/mspaint/atlimagedx.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+class CImageDx : public CImage
+{
+public:
+    CImageDx() : CImage()
+    {
+        GetImageHorizontalResolution = NULL;
+        GetImageVerticalResolution = NULL;
+        BitmapSetResolution = NULL;
+    }
+
+    BOOL GetResolution(Gdiplus::GpImage *pImage, float *pxDpi, float *pyDpi)
+    {
+        *pxDpi = 96;
+        *pyDpi = 96;
+
+        if (GetImageHorizontalResolution == NULL || GetImageVerticalResolution == NULL)
+        {
+            GetImageHorizontalResolution = AddrOf<GETIMAGEHORIZONTALRESOLUTION>("GdipGetImageHorizontalResolution");
+            GetImageVerticalResolution = AddrOf<GETIMAGEVERTICALRESOLUTION>("GdipGetImageVerticalResolution");
+        }
+
+        if (GetImageHorizontalResolution == NULL || GetImageVerticalResolution == NULL)
+            return FALSE;
+
+        GetImageHorizontalResolution(pImage, pxDpi);
+        GetImageVerticalResolution(pImage, pyDpi);
+        return TRUE;
+    }
+
+    BOOL SetResolution(Gdiplus::GpBitmap *pBitmap, float xDpi, float yDpi) const
+    {
+        if (xDpi <= 0)
+            xDpi = 96;
+        if (yDpi <= 0)
+            yDpi = 96;
+
+        if (BitmapSetResolution == NULL)
+            BitmapSetResolution = AddrOf<BITMAPSETRESOLUTION>("GdipBitmapSetResolution");
+
+        if (BitmapSetResolution == NULL)
+            return FALSE;
+
+        BitmapSetResolution(pBitmap, xDpi, yDpi);
+        return TRUE;
+    }
+
+    HRESULT LoadDx(LPCTSTR pszFileName, float *pxDpi, float *pyDpi) throw()
+    {
+        // convert the file name string into Unicode
+        CStringW pszNameW(pszFileName);
+
+        // create a GpBitmap object from file
```

Comment:
Why not just calling CImage::Load instead of seemingly redoing the whole work?

Reply:
`GpBitmap` is not available from outside of `CImage::Load`.